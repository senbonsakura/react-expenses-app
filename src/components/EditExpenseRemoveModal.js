import React from 'react';
import Modal from 'react-modal';

const RemoveModal = ({isOpen, onCloseModal, handleRemoveExpense, handleCloseModal}) => (
  <Modal
    isOpen={isOpen}
    onRequestClose={handleCloseModal}
    contentLabel="Remove Expense"
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className="modal__title">Are you sure to remove this expense?</h3>
    <div className="modal__body">
      <button className="button" onClick={handleRemoveExpense}>Yes</button>
      <button className="button button--secondary" onClick={handleCloseModal}>No</button>
    </div>
  </Modal>
);

export default RemoveModal;
