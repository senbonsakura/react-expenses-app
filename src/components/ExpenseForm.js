import { Field, Form, Formik } from 'formik';
import moment from 'moment';
import React, { Component } from 'react';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import * as Yup from 'yup';

//
// class ExpenseForm extends Component {
//   onDescriptionChange = (e) => {
//     const description = e.target.value;
//     this.setState(() => ({description}));
//   };
//   onNoteChange = (e) => {
//     const note = e.target.value;
//     this.setState(() => ({note}));
//   };
//   onAmountChange = (e) => {
//     const amount = e.target.value;
//     if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
//       this.setState(() => ({amount}));
//     }
//   };
//   onDateChange = (createdAt) => {
//     if (createdAt) {
//       this.setState(() => ({createdAt}));
//
//     }
//   };
//   onFocusChange = ({focused}) => {
//     this.setState(() => ({calendarFocused: focused}));
//   };
//   onSubmit = (e) => {
//     e.preventDefault();
//     if (!this.state.description || !this.state.amount) {
//       this.setState(() => ({
//         error: 'Please provide amount and description'
//       }));
//     } else {
//       this.setState(() => ({
//         error: ''
//       }));
//       this.props.onSubmit({
//         description: this.state.description,
//         amount: parseFloat(this.state.amount, 10) * 100,
//         createdAt: this.state.createdAt.valueOf(),
//         note: this.state.note
//       });
//     }
//   };
//
//   constructor (props) {
//     super(props);
//     this.state = {
//       description: props.expense ? props.expense.description : '',
//       note: props.expense ? props.expense.note : '',
//       amount: props.expense ? (props.expense.amount / 100).toString() : '',
//       createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
//       calendarFocused: false,
//       error: ''
//     };
//   }
//
//   render () {
//     return (
//       <form className="form" onSubmit={this.onSubmit}>
//         {this.state.error && <p className="form__error">{this.state.error}</p>}
//         <input
//           type="text"
//           placeholder="Description"
//           autoFocus
//           className="text-input"
//           value={this.state.description}
//           onChange={this.onDescriptionChange}/>
//         <input
//           type="text"
//           placeholder="Amount"
//           className="text-input"
//           value={this.state.amount}
//           onChange={this.onAmountChange}/>
//         <SingleDatePicker
//           date={this.state.createdAt}
//           onDateChange={this.onDateChange}
//           focused={this.state.calendarFocused}
//           onFocusChange={this.onFocusChange}
//           numberOfMonths={1}
//           isOutsideRange={(day) => false}
//           block={true}
//         />
//         <textarea
//           className="text-area"
//           placeholder="Add a note for your expense (optional)"
//           value={this.state.note}
//           onChange={this.onNoteChange}/>
//         <div>
//         <button className="button">Save Expense</button>
//         </div>
//       </form>
//
//     );
//   }
// }

class ExpenseForm extends Component {
  state = {
    calendarFocused: null
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  };

  render () {

    return (
      <div>
        <Formik
          validationSchema={Yup.object().shape({
            description: Yup.string().required('Please enter a description'),
            amount: Yup.string().required('Please enter an amount').matches(/^\d{1,}(\.\d{0,2})?$/, 'Please enter a valid number')
          })}
          initialValues={{
            description: this.props.expense ? this.props.expense.description : '',
            amount: this.props.expense ? this.props.expense.amount / 100 : '',
            note: this.props.expense ?this.props.expense.note : '',
            createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
          }}
          onSubmit={(values, {setFieldError, setSubmitting}) => {
            if (!values.description) {
              this.setState(() => {
                setFieldError('description', 'Please enter a description');
              });

            } else {

              this.props.onSubmit({
                description: values.description,
                amount: parseFloat(values.amount, 10) * 100,
                createdAt: values.createdAt.valueOf(),
                note: values.note
              });
            }

            setSubmitting(false);
          }}
          render = {({values, touched, errors, dirty, isSubmitting, handleChange})  => (
          <Form className="form">
            {touched.description && errors.description && <p className="form__error">{errors.description}</p>}

            <Field
              type="text"
              name="description"
              autoFocus
              className="text-input"
              placeholder="Description"
            />
            {touched.amount && errors.amount && <p className="form__error">{errors.amount}</p>}

            <Field
              type="text"
              name="amount"
              className="text-input"
              placeholder="Amount"
            />
            <SingleDatePicker
              date={values.createdAt}
              onDateChange={handleChange}
              focused={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              block={true}
            />
            <Field
              type="textarea"
              name="note"
              className="text-input"
              placeholder="Add a note for your expense (optional)"
            />

            <button disabled={isSubmitting} type="submit" className="button">Save Expense</button>


          </Form>
          )}
        >

        </Formik>
      </div>

    );
  }
}

// const FormikExpenseForm = withFormik({
// mapPropsToValues({ description, amount, note, createdAt}) {
//   return {
//     description: description || '',
//     amount: amount ? amount / 100 :0,
//     note: note || '',
//     createdAt: createdAt ? moment(moment) : moment(),
//   }
// },
//   validationSchema: Yup.object().shape({
//     description: Yup.string().required("Please enter a description"),
//     amount: Yup.string().required("Please enter an amount").matches(/^\d{1,}(\.\d{0,2})?$/,"Please enter a valid number"),
//   }),
//   handleSubmit(values , {setFieldError, setSubmitting}) {
//     if (!values.description) {
//       this.setState(() => {
//         setFieldError("description", 'Please enter a description')
//     });
//
//     } else {
//
//       this.props.onSubmit({
//         description: values.description,
//         amount: parseFloat(values.amount, 10) * 100,
//         createdAt: values.createdAt.valueOf(),
//         note: values.note
//       });
//     }
//
//     setSubmitting(false)
//   }
//
//
// })(ExpenseForm);

export default ExpenseForm;
