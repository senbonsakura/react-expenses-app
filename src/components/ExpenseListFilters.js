import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };
    onDatesChange = ({startDate, endDate}) => {
        this
            .props
            .setStartDate(startDate);
        this
            .props
            .setEndDate(endDate);
    };
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({calenderFocused}))
    };
    onSortChange = (e) => {
        if (e.target.value === 'amount') {
            this
                .props
                .sortByAmount()
        } else if (e.target.value === 'date') {
            this
                .props
                .sortByDate()
        }
    }
    onTextChange = (e) => {
        this
            .props
            .setTextFilter(e.target.value)
    }

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>

                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                    startDateId="start"
                    endDateId="end"

                />

            </div>
        )
    }

}

const mapStateToProps = (state) => ({filters: state.filters})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);