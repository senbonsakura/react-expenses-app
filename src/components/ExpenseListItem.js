import moment from 'moment';
import numeral from 'numeral';
import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({dispatch, description, amount, createdAt, id}) => (

  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item_title">{description}</h3>
      <span className="list-item__sub-title">{moment(createdAt).format('DD MMMM, YYYY')}</span>

    </div>
    <h3 className="list-item__data">
      {numeral(amount / 100).format('$0,0.00')}
    </h3>
  </Link>

);

export default ExpenseListItem;