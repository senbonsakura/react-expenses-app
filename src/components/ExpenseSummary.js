import React from 'react'
import numeral from 'numeral'
import getExpensesTotal from '../selectors/expenses-total'
import selectExpenses from '../selectors/expenses'
import { connect } from 'react-redux'

export const ExpenseSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
  const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,0.00')
  return (
    <div>
      <p>Viewing {expenseCount} {expenseWord} totaling {formattedExpenseTotal}</p>
    </div>
  )

}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses)
  }

}

export default connect(mapStateToProps)(ExpenseSummary)
