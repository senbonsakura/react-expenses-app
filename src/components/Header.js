import React from 'react'
import { Link } from 'react-router-dom'
import {connect} from 'react-redux'
import { startLogout } from '../actions/auth'
import Dropdown, {DropdownContent, DropdownTrigger} from 'react-simple-dropdown';


export const Header = ({startLogout, photoURL, displayName}) => (
  <header className="header" >
    <div className="content-container">
      <div className="header__content">
      <Link className="header__title" to="/dashboard">
        <h1>Expensify</h1>
      </Link>
        <div className="dropdown__container">


           <Dropdown>
             <DropdownTrigger>
               <img src={`${photoURL}`} className="header--avatar" />
               <span className="header--username">{displayName}</span>
               <i className="down" />
             </DropdownTrigger>
             <DropdownContent>
                   <button className="button buton--menu" onClick={startLogout}>Logout</button>
             </DropdownContent>
           </Dropdown>
          </div>
      </div>
    </div>
  </header>

);


const mapStateToProps = (state) => ({
  displayName: state.auth.displayName,
  photoURL: state.auth.photoURL,


});


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(mapStateToProps, mapDispatchToProps)(Header)
