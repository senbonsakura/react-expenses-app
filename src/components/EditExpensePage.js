import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import RemoveModal from './Modal';
import ExpenseForm from './ExpenseForm';
import Modal from 'react-modal';
import { closeDialog, openDialog } from 'redux-dialog';



export class EditExpensePage extends React.Component {
  state = {
    modalIsOpen: false
  };

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onCloseModal = () => {
    this.props.startCloseModal();
  };

  onRemove = () => {
    this.props.startRemoveExpense({id: this.props.expense.id});
    this.props.history.push('/');
  };

  componentWillMount() {
    Modal.setAppElement('body');
  }

  render () {
    return (

      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
          <button className="button button--secondary" onClick={this.props.startOpenModal}>
            Remove Expense
          </button>
        </div>
        <RemoveModal
          handleRemoveExpense={this.onRemove}
          onRequestClose={this.onCloseModal}
          className="modal"
        />
      </div>

    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state
    .expenses
    .find((expense) => props.match.params.id === expense.id)

});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
  startOpenModal: ()=> dispatch(openDialog("Remove Modal",{
    title:"Are you sure to remove this expense?",
  })),
  startCloseModal: ()=> dispatch(closeDialog("Remove Modal")),

});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
