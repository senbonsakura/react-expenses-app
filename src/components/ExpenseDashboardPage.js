import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary'
import LoginPage from './LoginPage'

const ExpenseDashboardPage = () => (
  <div>
    <LoginPage/>
    <ExpenseSummary />
    <ExpenseListFilters/>
    <ExpenseList/>
  </div>

)

export default ExpenseDashboardPage