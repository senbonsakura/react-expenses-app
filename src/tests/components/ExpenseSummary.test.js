import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';


test('should render ExpenseSummary with no expense', () => {
  const wrapper = shallow(<ExpenseSummary expenseCount={0} expensesTotal={0}/>)
  expect(wrapper).toMatchSnapshot()
});

test('should render ExpenseSummary with one expense', () => {
  const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={5540}/>)
  expect(wrapper).toMatchSnapshot()
});


test('should render ExpenseSummary with multiple expense', () => {
  const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={4563456556}/>)
  expect(wrapper).toMatchSnapshot()
});
