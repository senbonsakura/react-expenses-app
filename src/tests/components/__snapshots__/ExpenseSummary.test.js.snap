// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totaling 
      <span>
        $45,634,565.56
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with no expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totaling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totaling 
      <span>
        $55.40
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
